### PR TITLE
Fix for issue#2764

### LINKED ISSUE #2764: Setting attributes should have dtype checks
Our Topology system is essentially statically typed.  I recently had a test fail when you tried setting an atom name to an integer:  https://github.com/MDAnalysis/mdanalysis/pull/2755/commits/97e1782942a8262acfab2990460a129a15ed52bd

This would have previously worked as `dtype=object` (what we use for numpy arrays of python strings) will allow a Python integer into the array.  

We should have a `check_dtype` decorator around `TopologyAttr.set_atoms` (and residue/segment) to check that the supplied values are the correct type.

### PR DESCRIPTION
**Pull request by darmis007**
Fixes #2764

Changes made in this Pull Request:
 - 

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by orbeckst** [comment]
There are also conflicts – merge the current develop branch into your PR to update the code.

**Comment by darmis007** [comment]
@orbeckst sir, have made the changes which has solved the issue with the multiple test errors

**Comment by orbeckst** [comment]
I resolved the merge conflict and that will rerun the tests. But in the last tests I still saw some that failed. Please check your tests once they are done and if they fail, look carefully where they fail and try to fix this. If it's not clear why things fail, ask --- it also helps to copy and paste the failure output and directly link to the log of the failing run.

**Comment by darmis007** [comment]
> I resolved the merge conflict and that will rerun the tests. But in the last tests I still saw some that failed. Please check your tests once they are done and if they fail, look carefully where they fail and try to fix this. If it's not clear why things fail, ask --- it also helps to copy and paste the failure output and directly link to the log of the failing run.

Sure @orbeckst  sir will look into the failed tests once they are generated. I observed in the previous build some builds passed while some failed which looked to me as a version issue, will resolve them and get back to you with a commit at the earliest

**Comment by darmis007** [comment]
It seems to be a version issue, here is the [link](https://dev.azure.com/mdanalysis/mdanalysis/_build/results?buildId=1104&view=logs&j=4e749fca-3d34-5c5a-3ce5-4d1276511512&t=8661735f-b8db-5f8d-0b87-1042a707e273&l=321)

File: package/MDAnalysis/core/topologyattrs.py [hunk 1]
```diff
@@ -57,6 +57,35 @@
 from .. import _TOPOLOGY_ATTRS, _TOPOLOGY_TRANSPLANTS, _TOPOLOGY_ATTRNAMES
 
 
+def _check_dtype(func):
```
Comment (richardjgowers) [inline package/MDAnalysis/core/topologyattrs.py:60]: This wrapper is missing the point of the issue (maybe the issue is poorly worded). We need to make sure that TopologyAttrs with dtype object (eg AtomNames) don't get integers added to their value array. I.e. "1" (a string) is fine but 1 (an int, but still a python object) isn't.

**Review by richardjgowers — CHANGES_REQUESTED** [review]

**Comment by orbeckst** [comment]
I am closing this PR as stale. If you intend to pick it up again we can reopen it.

**Pull request closed without merging**